src/basic/namespace-util.c

```diff
@@ -210,14 +213,65 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
 
         /* Forks off a process in a new userns, configures the specified uidmap/gidmap, acquires an fd to it,
          * and then kills the process again. This way we have a userns fd that is not bound to any
-         * process. We can use that for file system mounts and similar. */
+         * process. We can use that for file system mounts and similar.
+         * If the owner of the new namespace is specified, setuid to it in the child. Use a pipe to sync and
+         * ensure setuid() happens first, then unshare(), then the mapping update. */
 
-        r = safe_fork("(sd-mkuserns)", FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|FORK_NEW_USERNS, &pid);
+        if (pipe2(pipe_fd, 0) < 0)
+                return log_error_errno(errno, "Failed to open bidirectional pipe: %m");
+
+        /* If we have to set a different owner we have to setuid before unsharing the new namespace, otherwise
+         * the kernel will refuse the operation, so don't fork directly in the new user namespace in that case.
+         * If we don't have to change the owner then it easier and we can let the clone set up the namespace
+         * instead via the apposite flag. */
+        if (owner_uid == UID_INVALID)
```

Comment:
shouldn't this be `owner_uid == UID_INVALID || owner_gid == GID_INVALID`?

Reply:
I just figured you'd either change none, or just the uid, or both, but just the gid seems strange as a use case. But can change it